src/canard.c

```diff
@@ -1,7 +1,919 @@
+                                     uint16_t bit_offset,
+                                     uint8_t bit_length)
+{
+    uint64_t bits = 0;
+    int shift_val =  bit_length - 8;
+    if (transfer->payload_len > 7)      //multi frame
+    {
+        CanardBufferBlock* block = transfer->payload_middle;
+        int i;
+        uint8_t index = 0;
+
+        if (CANARD_RX_PAYLOAD_HEAD_SIZE > 0)    //head
+        {
+            for (i = 0; i<CANARD_RX_PAYLOAD_HEAD_SIZE && index<bit_length && shift_val>=0; i++, index++)
+            {
+                if (index>=bit_offset / 8)
+                {
+                    bits |= ((uint64_t)transfer->payload_head[i] << shift_val);
+                    shift_val -= 8;
+                }
+            }
+        }
+        //middle (buffer blocks)
+        for (i = 0; index<(CANARD_RX_PAYLOAD_HEAD_SIZE + transfer->middle_len) &&
+             index<bit_length && shift_val>=0; i++, index++)
+        {
+            if (index>=bit_offset / 8)
+            {
+                bits |= ((uint64_t)block->data[i] << shift_val);
+                shift_val -= 8;
+            }
+            if (i==CANARD_BUFFER_BLOCK_DATA_SIZE - 1)
+            {
+                i = 0;
+                block = block->next;
+            }
+        }
+        // tail
+        int tail_len = transfer->payload_len - (CANARD_RX_PAYLOAD_HEAD_SIZE + transfer->middle_len);
+        for (i = 0; i<(tail_len) && index<bit_length && shift_val>=0; i++, index++)
+        {
+            if (index>=bit_offset / 8)
+            {
+                bits |= ((uint64_t)transfer->payload_tail[i] << shift_val);
+                shift_val -= 8;
+            }
+        }
+    }
+    else    //single frame
+    {
+        uint8_t i;
+        for (i = 0; i<transfer->payload_len && i<bit_length && shift_val>=0; i++)
+        {
+            if (i>=bit_offset / 8)
+            {
+                bits |= ((uint64_t)transfer->payload_head[i] << shift_val);
+                shift_val -= 8;
+            }
+        }
+    }
+    return bits;
+}
+
+/**
+ * This function can be invoked by the application to release pool blocks that are used
+ * to store the payload of this transfer
+ */
+uint64_t canardReleaseRxTransferPayload(CanardInstance* ins, CanardRxTransfer* transfer)
+{
+    CanardBufferBlock* temp = transfer->payload_middle;
+    while (transfer->payload_middle != NULL)
+    {
+        temp = transfer->payload_middle->next;
+        canardFreeBlock(&ins->allocator, transfer->payload_middle);
+        transfer->payload_middle = temp;
+    }
+    transfer->payload_middle = NULL;
+
+    transfer->payload_head = NULL;
+    transfer->payload_tail = NULL;
+    transfer->payload_len = 0;
+    return 0;
+}
+
+/**
+ * This function can be invoked by the application to release pool blocks that are used
+ * to store the payload of this transfer
+ */
+CANARD_INTERNAL uint64_t canardReleaseStatePayload(CanardInstance* ins, CanardRxState* rxstate)
```

Comment:
Wait, it is. Sorry, confused it with `canardReleaseRxTransferPayload`. It should go below, under `internal (static functions)`.